SmartPointers/shared_pointer.hpp

```diff
@@ -1,80 +1,134 @@
-        {
+    // Copy Constructor
+    SharedPointer(const SharedPointer& p) : m_ptr(p.m_ptr), count(p.count) { 
+        std::cout << "called  Shared Copy Constructor \n";
+        (*count)++; 
         }
+    
+    // Destructor
+    ~SharedPointer() { 
+        std::cout << " Shared Destructor \n";
+        destroy(); }
 
-    SharedPointer(const SharedPointer& sharedPtr){ // Copy constructor
-        this->m_ptr = sharedPtr.m_ptr;
-        this->m_count = sharedPtr.m_count;
-        if(nullptr != sharedPtr.m_ptr)
+    // Destroy the object
+    void destroy() {
+        std::cout << " Shared Destroy Called \n";
+        if(count == nullptr)
         {
-            (*this->m_count)++;
+            return;
+        }
+        *count = *count > 0 ? *count - 1 : *count;
+        if (*count == 0) {
+            delete m_ptr;
+            delete count;
         }
-        
     }
-    
-    SharedPointer& operator=(const SharedPointer& sharedPtr) //Copy assignment operator
-    {
-        this->m_ptr = sharedPtr.m_ptr;
-        this->m_count = sharedPtr.m_count;
-        if(nullptr != sharedPtr.m_ptr)
-        {
-            (*this->m_count)++;
+
+    // Copy assigment operator
+    SharedPointer& operator=(const SharedPointer& p) {
+        std::cout << " Copy Assigment Operator Called \n";
+        if (this != &p) {
+            destroy();
+            m_ptr = p.m_ptr;
+            count = p.count;
+            (*count)++;
         }
+        return *this;
     }
-    
-    T* operator->() const
-    {
-        return this->m_ptr;
+
+    // Move Constructor
+    SharedPointer(SharedPointer&& p) : m_ptr(p.m_ptr), count(p.count) {
+        std::cout << "Move Shared Constructor \n";
+        p.m_ptr = nullptr;
+        p.count = nullptr;
     }
-    
 
-     void cleanup(){
-            cout << "Cleanup is called";
-            (*m_count)--;
-            if (*m_count == 0)
-            {
-                if (nullptr != m_ptr)
-                    delete m_ptr;
-                delete m_count;
+    // Move Assigment operator
+    SharedPointer operator=(SharedPointer&& p) {
+        std::cout << "Move Shared Assigment Operator \n";
+        if (this != &p) {
+            destroy();
+            m_ptr   = p.m_ptr;
+            count   = p.count;
+            p.m_ptr = nullptr;
+            p.count = nullptr;
+        }
+        return *this;
+    }
+
+    // Dereferencing
+    T& operator*() { return *m_ptr; }
+
+    // Pointing
+    T* operator->() { return m_ptr; }
+
+    // Return use_count
+    int use_count() const { return *count; }
+
+    // calculate unique
+    bool unique() const { return *count == 1; }
+
+    // Reset
+    void reset() {
+        destroy();
+        m_ptr = nullptr;
+        count = new int(1);
+    }
+
+    // Paramatized reset
+    void reset(SharedPointer& new_ptr) {
+        destroy();
+        m_ptr = new_ptr.m_ptr;
+        count = new_ptr.count;
+        *count++;
+    }
+
+    // release
+    T* release() {
```

Comment:
Shared ptr don't have release interface ( Doesn't make sense for shared ptr to release ownership to row ptr, because what will happen to those other objects sharing ownership ? and what is released ptr is manually deleted ? )

https://cplusplus.com/reference/memory/shared_ptr/

Reply:
done